sdk/core/azure-core/src/base64.cpp

```diff
@@ -4,93 +4,474 @@
+  i0 = Base64DecodeArray[i0];
+  i1 = Base64DecodeArray[i1];
+  i2 = Base64DecodeArray[i2];
+  i3 = Base64DecodeArray[i3];
+
+  i0 <<= 18;
+  i1 <<= 12;
+  i2 <<= 6;
+
+  i0 |= i3;
+  i1 |= i2;
+
+  i0 |= i1;
+  return i0;
+}
+
+static void base64WriteThreeLowOrderBytes(std::vector<uint8_t>::iterator destination, int64_t value)
+{
+  *destination = static_cast<uint8_t>(value >> 16);
+  *(destination + 1) = static_cast<uint8_t>(value >> 8);
+  *(destination + 2) = static_cast<uint8_t>(value);
+}
+
+std::vector<uint8_t> base64Decode(const std::string& text)
+{
+
+  if (text.size() < 4)
   {
-    std::string encoded;
-    // According to RFC 4648, the encoded length should be ceiling(n / 3) * 4
-    DWORD encodedLength = static_cast<DWORD>((data.size() + 2) / 3 * 4);
-    encoded.resize(encodedLength);
-
-    CryptBinaryToStringA(
-        reinterpret_cast<const BYTE*>(data.data()),
-        static_cast<DWORD>(data.size()),
-        CRYPT_STRING_BASE64 | CRYPT_STRING_NOCRLF,
-        static_cast<LPSTR>(&encoded[0]),
-        &encodedLength);
-
-    return encoded;
+    return std::vector<uint8_t>(0);
   }
 
-  std::vector<uint8_t> Convert::Base64Decode(const std::string& text)
+  int64_t sourceIndex = 0;
+  int64_t destinationIndex = 0;
+  auto inputSize = text.size();
+  auto inputPtr = text.data();
+  // use the size for the max decoded size
+  auto maxDecodedSize = (inputSize / 4) * 3;
+  std::vector<uint8_t> destination(maxDecodedSize);
+  auto destinationPtr = destination.begin();
+
+  while (sourceIndex < static_cast<int64_t>(inputSize) - 4)
+  {
+    int64_t result = base64Decode(inputPtr + sourceIndex);
+    base64WriteThreeLowOrderBytes(destinationPtr, result);
+    destinationPtr += 3;
+    destinationIndex += 3;
+    sourceIndex += 4;
+  }
+
+  // We are guaranteed to have an input with at least 4 bytes at this point, with a size that is a
+  // multiple of 4.
+  int64_t i0 = *(inputPtr + inputSize - 4);
+  int64_t i1 = *(inputPtr + inputSize - 3);
+  int64_t i2 = *(inputPtr + inputSize - 2);
+  int64_t i3 = *(inputPtr + inputSize - 1);
+
+  i0 = Base64DecodeArray[i0];
+  i1 = Base64DecodeArray[i1];
+
+  i0 <<= 18;
+  i1 <<= 12;
+
+  i0 |= i1;
+
+  if (i3 != EncodingPath)
   {
-    std::vector<uint8_t> decoded;
-    // According to RFC 4648, the encoded length should be ceiling(n / 3) * 4, so we can infer an
-    // upper bound here
-    DWORD decodedLength = DWORD(text.length() / 4 * 3);
-    decoded.resize(decodedLength);
-
-    CryptStringToBinaryA(
-        text.data(),
-        static_cast<DWORD>(text.length()),
-        CRYPT_STRING_BASE64 | CRYPT_STRING_STRICT,
-        reinterpret_cast<BYTE*>(decoded.data()),
-        &decodedLength,
-        nullptr,
-        nullptr);
-    decoded.resize(decodedLength);
-    return decoded;
+    i2 = Base64DecodeArray[i2];
+    i3 = Base64DecodeArray[i3];
+
+    i2 <<= 6;
+
+    i0 |= i3;
+    i0 |= i2;
+
+    base64WriteThreeLowOrderBytes(destinationPtr, i0);
```

Comment:
Same here and elsewhere - add input validation checks (if i0 < 0) and throw for invalid input

https://github.com/Azure/azure-sdk-for-c/blob/d7b388c68b07ee769dd9b061114b93336fd53385/sdk/src/azure/core/az_base64.c#L487-L490